haskell/experimental/private/module.bzl

```diff
@@ -214,40 +196,117 @@ def haskell_module_impl(ctx):
                 plugin_dep_info.hs_libraries,
                 plugin_tool_inputs,
                 preprocessors_inputs,
-            ] + [
                 # TODO[AH] Factor this out
                 # TODO[AH] Include object files for template Haskell dependencies.
-                depset(direct = [dep[HaskellModuleInfo].interface_file])
-                for dep in ctx.attr.deps
-                if HaskellModuleInfo in dep
+                interface_inputs,
             ],
         ),
         input_manifests = preprocessors_input_manifests + plugin_tool_input_manifests,
-        outputs = [obj, interface],
+        outputs = objs + interfaces,
         mnemonic = "HaskellBuildObject" + ("Prof" if with_profiling else ""),
-        progress_message = "HaskellBuildObject {}".format(hs.label),
+        progress_message = "HaskellBuildObject {} {}".format(hs.label, dep.label),
         env = hs.env,
         arguments = args,
+        extra_name = dep.label.package.replace("/", "_") + "_" + dep.label.name,
     )
 
-    # Construct the import search paths for this module
-    workspace_root = paths.join(ctx.bin_dir.path, ctx.label.workspace_root)
-    package_root = paths.join(workspace_root, ctx.label.package)
-    src_strip_prefix = ctx.attr.src_strip_prefix
+def get_module_path_from_target(dep):
+    src = dep[HaskellModuleInfo].attr.src.files.to_list()[0].path
+    src_strip_prefix = dep[HaskellModuleInfo].attr.src_strip_prefix
+
     if src_strip_prefix.startswith("/"):
-        import_dir = paths.join(workspace_root, src_strip_prefix[1:])
+        prefix_path = paths.join(src_strip_prefix[1:])
```

Comment:
I think the `workspace_root` prefix is relevant only for paths of generated files. On the contrary, `prefix_path` here is used to strip a prefix from the path of a source file.

Reply:
I think one place where this is relevant is for targets built in external workspaces.
Does keeping this cause errors? If not, I'd keep it, or at least verify that a target in an external workspace still works correctly without this.

Reply:
It does cause errors because `paths.relativize` will complain if `prefix_path` is not actually a prefix.

Is there a way to determine whether the target belongs to an external workspace?

Reply:
I think if [`workspace_root`](https://docs.bazel.build/versions/main/skylark/lib/Label.html#workspace_root) is empty then it's in the main workspace.  That said, it's probably a good idea to have an external workspace test case for a `haskell_module` in general. There are some subtleties around these. We have a similar test case [here](https://github.com/tweag/rules_haskell/tree/master/tests/external-haskell-repository).

Reply:
It is my confusion. I added `workspace_root` back. Adding tests is still pending.

Reply:
Thanks for looking into this!

Reply:
I added a test.